src/test/test_ep_cred_client.c

```diff
@@ -140,16 +152,42 @@ test_run()
 		assert(rc == 0);
 
 		input = crt_req_get(rpc);
-		if (i == 0)
-			input->pi_delay = 1;
-		else
+
+		if (i == 0) {
+			/* If we test 'send to front of queue' flag we
+			 * want to increase response delay of first rpc to 3
+			 * seconds in order to allow sufficient queue up
+			 */
+			if (test.tg_send_queue_front)
+				input->pi_delay = 3;
+			else
+				input->pi_delay = 1;
+		} else {
 			input->pi_delay = 0;
+		}
 
 		rc = crt_req_send(rpc, rpc_handle_reply, NULL);
 		assert(rc == 0);
 		sent_count++;
 	}
 
+	/* Send RPC message to be in front of the queue. This option
+	 * should only be used when test.tg_burst_count is large while
+	 * test.tg_credits is small, allowing sufficient queue up of
+	 * rpcs
+	 */
+	if (test.tg_send_queue_front) {
+		rc = crt_req_create(test.tg_crt_ctx, &ep, OPC_PING_FRONT,
+				&rpc);
+		assert(rc == 0);
+
+		rc = crt_req_send(rpc, rpc_handle_ping_front_q, NULL);
+		assert(rc == 0);
+
+		test_sem_timedwait(&test.tg_front_queue_token, 61, __LINE__);
+		assert(sent_count != resp_count);
```

Comment:
why not move this assert to line 132?  you don't need the semaphore if you do that.

Reply:
The idea here is that when we get response back from RPC that was sent to front queue, we want to ensure that other rpcs (which were submitted before that) havent finished; so basically that pushing RPC to the front of the queue actually works

Reply:
moving the assert to line 132 has the same effect without the use of a semaphore, no?

Reply:
Not quite, its possible (due to a bug) to never get a response from this rpc. So for that reason we do need a timed semaphore -- report errors if this queued rpc never got responded to.